objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java

```diff
@@ -355,28 +356,40 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
            return;
        }
        
+       Map<String, Object> processingParams = new LinkedHashMap<>();
+       processingParams.put("timestamp", imageResults.getTimestamp());
+       
        // Send the results of the neural net, or send a Query error if an exception occurs
+       NeuralNetInterface localNeuralNet = null;
        try {
-           synchronized (this) {
+           synchronized (SYNCH_KEY) {
                if (neuralNet == null) { // Should only happen when close() runs just before sendDataFromImage()
                    if (client != null) {
                        client.sendQueryError(replyAddress, this.getClass().getPackage().getName() + ".closed",
                                "The source closed mid message", null);
                    }
                    return;
                }
-               NeuralNetResults results = neuralNet.processImage(image, request);
-               lastQueryFilename = results.getLastFilename();
-               
-               // Send the normal message as the response if requested, otherwise just send the data 
-               if (request.get("sendFullResponse") instanceof Boolean && (Boolean) request.get("sendFullResponse")) {
-                   Map response = createMapFromResults(imageResults, results);
-                   client.sendQueryResponse(200, replyAddress, response);
-               } else if (results.getResults().isEmpty()) {
-                   client.sendQueryResponse(204, replyAddress, new LinkedHashMap<>());
-               } else {
-                   client.sendQueryResponse(200, replyAddress, results.getResults().toArray(new Map[0]));
-               }
+               localNeuralNet = neuralNet;
+           }
+           
+           NeuralNetResults results;
+           if (localNeuralNet instanceof NeuralNetInterface2) {
+               NeuralNetInterface2 localNeuralNet2 = (NeuralNetInterface2) localNeuralNet;
+               results = localNeuralNet2.processImage(processingParams, image, request);
```

Comment:
here & above, since the local variable is used only on this line, casting is probably a better choice.  The compiler probably figured that out and did it anyway, but clearer what's going on.